sunrise.py tests

nsryan2 opened the pull request:
This PR adds tests for all of the functions in sunrise.py to the tests folder. Of the 29 tests, currently 9 of them fail. The differences between the calculations and expectations are small and likely due to slightly different models or approximations. Several of the tests make use of an [excel spreadsheet](https://www.esrl.noaa.gov/gmd/grad/solcalc/NOAA_Solar_Calculations_day.xls) published by NOAA which does a very detailed analysis based off of a given utc, longitude, latitude, and date. All nine tests that fail have expected values from this spreadsheet. If there's anything I can improve, please let me know.

nsryan2 committed:
This adds the tests for all of the functions in sunrise.py

nsryan2 committed:
Merge branch 'iss26-sunrise' into nsryan-iss26

samgdotson commented:
Don't forget to assign me as a reviewer!

samgdotson reviewed (changes requested):
@nsryan2 The tests should all pass before pulling into the repo.
I'm really glad that you're comparing to expected values! That's excellent! However, it appears that the error might be coming from a difference in precision. My implementation is indeed an approximation, which is okay! 
There is a pytest function called ``approx`` that allows you to compare to an approximate value. I would recommend changing it to this and make sure the tests pass!
https://docs.pytest.org/en/stable/reference.html#id12

nsryan2 commented:
I'll look into that, thanks for the suggestion! I think because this is your personal repository I am not able to add an assignee or reviewer. I looked to do that when I made the PR but it wouldn't let me, if you go into the settings on the repository I think you can change this under interaction limits (or you could add me in some other way). I'm not too sure, but I ran into a similar problem when I was showing a friend how to use Github. 

How should I set the tolerance on the approximations?

samgdotson commented:
Ah it seems you're right about the limits (since I am not an organization). You can @me if we have to do this again. Although in the future you should be able to make direct PRs to CAIRO. 

As far as tolerances... they should be as precise as possible, but no more. I know it's an approximation, and it doesn't have to be exact. I think at worst (the least precise) would be +/- 0.1

Thanks Nathan!

nsryan2 committed:
Add approx to everything except equation of time

The difference between the observed and expected is around 1.56 for all four equation of time tests

samgdotson commented on tests/test_sunrise.py:
```diff
@@ -1,8 +1,371 @@
-from sunrise import frac_year
+from sunrise import *
 import numpy as np
+import datetime as dt
+import matplotlib.pyplot as plt
```
```suggestion
```

These packages aren't needed to run the tests

samgdotson commented on tests/test_sunrise.py:
```diff
@@ -1,8 +1,371 @@
-from sunrise import frac_year
+from sunrise import *
```
```suggestion
from sunrise import *
from pytest import approx
```
We need to import approx from pytest

samgdotson commented on tests/test_sunrise.py:
```diff
@@ -1,8 +1,371 @@
+    was used.
+    """
+    obs_i = -3.18827461
+    exp_i = equation_of_time(0.1, True)
```
``equation_of_time`` doesn't have an argument for leap years. Which probably should be added in ``sunrise.py``.

samgdotson replied:
In ``equation_of_time`` add the argument ``leap_year=False`` and then add it to the ``frac_year`` call

```python
B = frac_year(hour, leap_year)
```

samgdotson reviewed (changes requested):
After making some changes locally, 4 tests still fail. All related to the ``equation_of_time`` function.

The original ET I used was from [this website](https://www.pveducation.org/pvcdrom/properties-of-sunlight/solar-time#HRA). I updated the approximation to match [this other website](https://www.intmath.com/blog/mathematics/the-equation-of-time-5039) and the agreement seems to improve slightly.
The new ET is
```python
et = -7.655*np.sin(B*np.pi/180) + 9.873*np.sin((2*B+3.588)*np.pi/180)
```
Keep up the good work, we're almost there. 

**Before continuing, do this**
```bash
git pull https://github.com/samgdotson/cairo.git main
```
Then you will be up-to-date

nsryan2 committed:
Apply suggestions from code review

Co-authored-by: Sam Dotson <44342873+samgdotson@users.noreply.github.com>

nsryan2 commented:
> ```python
> et = -7.655*np.sin(B*np.pi/180) + 9.873*np.sin((2*B+3.588)*np.pi/180)
> ```

I switched to the new et, and I agree they're much closer, but the sign is wrong on the values. I was getting a positive number where there was expected to be a negative one, which is strange because the plot in the second site you referenced has plots that start off negative. Is that something you noticed as well? 

These tests fail for me
`FAILED test_sunrise.py::test_equation_of_time_initial - assert -3.421699572 == 3.54942053247605 ± 1.0e+00`
`FAILED test_sunrise.py::test_equation_of_time_later - assert -3.305115384 == 3.5494205324760584 ± 5.5e+00`
`FAILED test_sunrise.py::test_equation_of_time_initial_leap_year - assert -3.18827461 == 3.475543982650941 ± 5.4e+00`
`FAILED test_sunrise.py::test_equation_of_time_later_leap_year - assert -3.543467339 == 3.47554398265095 ± 5.4e+00`
`FAILED test_sunrise.py::test_time_correction_initial - assert -57.828798884565636 == -49.28174332688438`

samgdotson commented:
@nsryan2 Thanks for checking and testing this, there is a lesson here: We can't get rid of a test because it's failing. Clearly there is a bug in the code I wrote. I found the bug, and as your code indicated, it is indeed in the ``equation_of_time`` function. 
I walked through the function again and found that _in the original implementation_ I used a ``cos`` where I should use a ``sin``. I have now fixed this and the EoT im getting is 

```bash
The time is: 0
The hour number is: 0
The equation of time is: -3.256234923805355
The local standard time meridian is: -75
The time correction factor is: -56.23234292380536
The local solar time is: -0.937205715396756
The hour angle is: -194.05808573095135
The elevation angle is: -69.25943477025157
```
There is a second bug that seems to be causing the problem: The UTC shift for Champaign/Chicago is -6, not -5. 
Once I changed it to -6, everything works out and we get the correct sunset/sunrise times.

I've updated the main branch with these changes. Pull them down and let me know if they work. You can tighten up the approximations as well. They should be accurate to within 0.5 minutes.

```bash
The time is: 0
The hour number is: 0
The equation of time is: -3.2562349238053487
The local standard time meridian is: -90
The time correction factor is: 3.7676570761946406
The local solar time is: 0.06279428460324402
The hour angle is: -179.05808573095135
The elevation angle is: -72.97564456089731
Sunrise: 7.3378681491198625
```

**Before continuing, do this**
```bash
git pull https://github.com/samgdotson/cairo.git iss26-sunrise
```

nsryan2 committed:
Update tests with the changes to sunrise.py

Change all of the numbers affected by the change in UTC, the slightly different coordinates, and adding appropriate approximations to make sure all tests pass reasonably

nsryan2 commented:
@samgdotson Alright, so this most recent commit should fix the discontinuities with the UTC and formula change. All of the tests pass, and those that require an approximation are less than 0.1 (with the exception of the two listed below). Something I noticed about the branch iss26-sunrise is that it has n_days = 365 for a leap year in the frac_year equation. This PR could change that (it is listed as 366 on my version), but I was just wondering if that was intentional like in the day_number equation where 365 is in the equation itself.
test_time_correction_later_leap_year approx = 0.2
test_time_correction_initial approx = 0.183

samgdotson commented:
When you pull changes down, don't forget to push them to your branch!
I pulled your branch and the tests still fail because some of the functions have been renamed. Which you did in the tests, but it's not reflected in ``sunrise.py``.

There should be some changes to ``sunrise.py``
1. Equation_of_time should have 2 arguments
2. generate_time_series should be generate_elevation_series

Before I can merge this request you need to do the following:
```bash
git checkout nsryan-iss26
git pull https://github.com/samgdotson/cairo.git iss26-sunrise
git commit -am "makes update"
git push origin nsryan-iss26
```
**_you should not have to make any changes manually._**

samgdotson commented:
Good job @nsryan2! Thanks for the contribution!
